File: packages/earthsci_toolkit/src/earthsci_toolkit/discretize.py
"""Discretization pipeline per discretization RFC §11 (gt-gbs2).

The public entry point is :func:`discretize`, which walks a parsed ESM
document and emits a discretized ESM:

  1. Canonicalize all expressions (§5.4).
  2. Resolve model-level boundary conditions into a synthetic ``bc`` op
     so they flow through the same rule engine as interior equations.
  3. Apply the rule engine (§5.2) to every equation RHS and every BC
     value with a max-pass budget.
  4. Re-canonicalize the rewritten ASTs.
  5. Check for unrewritten PDE ops (§11 Step 7) — error or
     passthrough-annotate depending on ``strict_unrewritten``.
  6. Record ``metadata.discretized_from`` provenance.

Scheme-expansion of ``use:<scheme>`` rules (§7.2.1), cross-grid
``regrid`` wrapping, and the RFC §12 DAE binding contract are out of
scope here.

Mirrors ``packages/EarthSciSerialization.jl/src/discretize.jl`` at
commit ``5849c525`` (gt-gbs2). Acceptance criterion is parity with the
Julia reference on the shared conformance fixtures.
"""

from __future__ import annotations

import copy
from typing import Any, Callable, Dict, List, Mapping, Optional, Sequence

from .canonicalize import canonicalize
from .esm_types import Expr, ExprNode
from .parse import _parse_expression
from .rule_engine import (
    Rule,
    RuleContext,
    RuleEngineError,
    parse_rules,
    rewrite,
)
from .serialize import _serialize_expression


_PDE_OPS = frozenset({"grad", "div", "laplacian", "D", "bc"})


def discretize(
    esm: Mapping[str, Any],
    *,
    max_passes: int = 32,
    strict_unrewritten: bool = True,
) -> Dict[str, Any]:
    """Run the RFC §11 discretization pipeline on an ESM document.

    ``esm`` is the parsed ESM payload as a mapping (the form produced by
    decoding JSON with the stdlib). The function returns a new dict; the
    input is not mutated.

    Parameters
    ----------
    esm:
        The ESM document to discretize.
    max_passes:
        Per-expression rule-engine budget (§5.2.5). Default 32.
    strict_unrewritten:
        When ``True`` (default), a rewritten expression that still
        carries a PDE op (``grad``, ``div``, ``laplacian``, ``D``,
        ``bc``) raises :class:`RuleEngineError` with code
        ``E_UNREWRITTEN_PDE_OP``. When ``False``, the offending
        equation or BC is instead marked ``passthrough: true`` and the
        rewritten form is retained verbatim.

    Returns
    -------
    dict
        The discretized ESM. ``metadata.discretized_from`` is set to a
        provenance sub-object carrying the input's ``metadata.name``;
        the ``"discretized"`` tag is appended to ``metadata.tags``.
    """
    if not isinstance(esm, Mapping):
        raise TypeError(
            f"discretize: input must be a JSON object / mapping; got {type(esm).__name__}"
        )

    out: Dict[str, Any] = copy.deepcopy(dict(esm))

    top_rules = _load_rules(out.get("rules"))
    ctx = _build_rule_context(out)

    models = out.get("models")
    if isinstance(models, Mapping):
        for mname, mraw in list(models.items()):
            model = dict(mraw) if isinstance(mraw, Mapping) else mraw
            if isinstance(model, dict):
                _discretize_model(
                    str(mname), model, top_rules, ctx, max_passes, strict_unrewritten
                )
                models[mname] = model  # type: ignore[index]

    _record_discretized_from(out)
    return out


# ============================================================================
# Rule-context assembly (grids + variables)
# ============================================================================


def _build_rule_context(esm: Mapping[str, Any]) -> RuleContext:
    grids: Dict[str, Dict[str, Any]] = {}
    grids_raw = esm.get("grids")
    if isinstance(grids_raw, Mapping):
        for gname, graw in grids_raw.items():
            grids[str(gname)] = _extract_grid_meta(graw)

    variables: Dict[str, Dict[str, Any]] = {}
    models = esm.get("models")
    if isinstance(models, Mapping):
        for _mname, mraw in models.items():
            if not isinstance(mraw, Mapping):
                continue
            mgrid = mraw.get("grid")
            mgrid_str = str(mgrid) if isinstance(mgrid, str) else None
            vars_raw = mraw.get("variables")
            if not isinstance(vars_raw, Mapping):
                continue
            for vname, vraw in vars_raw.items():
                if not isinstance(vraw, Mapping):
                    continue
                meta: Dict[str, Any] = {}
                if mgrid_str is not None:
                    meta["grid"] = mgrid_str
                shape = vraw.get("shape")
                if shape is not None:
                    meta["shape"] = shape
                loc = vraw.get("location")
                if loc is not None:
                    meta["location"] = str(loc)
                variables[str(vname)] = meta
    return RuleContext(grids=grids, variables=variables)


def _extract_grid_meta(graw: Any) -> Dict[str, Any]:
    meta: Dict[str, Any] = {}
    if not isinstance(graw, Mapping):
        return meta
    dims_raw = graw.get("dimensions")
    if isinstance(dims_raw, Sequence) and not isinstance(dims_raw, (str, bytes)):
        spatial: List[str] = []
        periodic: List[str] = []
        nonuniform: List[str] = []
        for d in dims_raw:
            if not isinstance(d, Mapping):
                continue
            name = d.get("name")
            if name is None:
                continue
            name_s = str(name)
            spatial.append(name_s)
            if d.get("periodic") is True:
                periodic.append(name_s)
            spacing = d.get("spacing")
            if isinstance(spacing, str) and spacing in ("nonuniform", "stretched"):
                nonuniform.append(name_s)
        meta["spatial_dims"] = spatial
        meta["periodic_dims"] = periodic
        meta["nonuniform_dims"] = nonuniform
    return meta


# ============================================================================
# Model-level pipeline
# ============================================================================


def _discretize_model(
    mname: str,
    model: Dict[str, Any],
    top_rules: List[Rule],
    ctx: RuleContext,
    max_passes: int,
    strict_unrewritten: bool,
) -> None:
    local_rules_raw = model.get("rules")
    local_rules = _load_rules(local_rules_raw)
    rules = top_rules if not local_rules else list(top_rules) + list(local_rules)

    mp = _lookup_max_passes(model, max_passes)

    eqns = model.get("equations")
    if isinstance(eqns, list):
        for i, eqn_any in enumerate(eqns):
            if not isinstance(eqn_any, Mapping):
                continue
            eqn = dict(eqn_any)
            _discretize_equation(
                f"models.{mname}.equations[{i}]",
                eqn,
                rules,
                ctx,
                mp,
                strict_unrewritten,
            )
            eqns[i] = eqn

    bcs = model.get("boundary_conditions")
    if isinstance(bcs, Mapping):
        for bc_name, bc_any in list(bcs.items()):
            if not isinstance(bc_any, Mapping):
                continue
            bc = dict(bc_any)
            _discretize_bc(
                f"models.{mname}.boundary_conditions.{bc_name}",
                bc,
                rules,
                ctx,
                mp,
                strict_unrewritten,
            )
            bcs[bc_name] = bc  # type: ignore[index]


def _lookup_max_passes(model: Mapping[str, Any], default: int) -> int:
    rules_meta = model.get("rules_config")
    if isinstance(rules_meta, Mapping):
        mp = rules_meta.get("max_passes")
        if isinstance(mp, int) and not isinstance(mp, bool):
            return mp
    return default


# ============================================================================
# Per-equation / per-BC rewrite
# ============================================================================


def _discretize_equation(
    path: str,
    eqn: Dict[str, Any],
    rules: Sequence[Rule],
    ctx: RuleContext,
    max_passes: int,
    strict_unrewritten: bool,
) -> None:
    passthrough = _as_bool(eqn.get("passthrough", False))

    def _set_passthrough(v: bool) -> None:
        eqn["passthrough"] = v

    if "rhs" in eqn:
        eqn["rhs"] = _rewrite_or_passthrough(
            f"{path}.rhs",
            eqn["rhs"],
            rules,
            ctx,
            max_passes,
            strict_unrewritten,
            passthrough,
            _set_passthrough,
        )
    if "lhs" in eqn:
        eqn["lhs"] = _canonicalize_value(eqn["lhs"])


def _discretize_bc(
    path: str,
    bc: Dict[str, Any],
    rules: Sequence[Rule],
    ctx: RuleContext,
    max_passes: int,
    strict_unrewritten: bool,
) -> None:
    passthrough = _as_bool(bc.get("passthrough", False))

    def _set_passthrough(v: bool) -> None:
        bc["passthrough"] = v

    variable = bc.get("variable")
    kind = bc.get("kind")
    side = bc.get("side")
    value_raw = bc.get("value")

    rewritten_via_bc_rule = False
    if (
        isinstance(variable, str)
        and isinstance(kind, str)
        and rules
    ):
        wrapper: Dict[str, Any] = {
            "op": "bc",
            "args": [variable],
            "kind": kind,
        }
        if isinstance(side, str):
            wrapper["side"] = side
        if value_raw is not None:
            wrapper["args"].append(value_raw)
        bc_expr = _parse_expression(wrapper)
        rewrite_out = rewrite(canonicalize(bc_expr), rules, ctx, max_passes=max_passes)
        if not (isinstance(rewrite_out, ExprNode) and rewrite_out.op == "bc"):
            final = canonicalize(rewrite_out)
            if _has_pde_op(final) and not passthrough:
                if strict_unrewritten:
                    op = _first_pde_op(final)
                    raise RuleEngineError(
                        "E_UNREWRITTEN_PDE_OP",
                        f"{path}.value still contains PDE op '{op}' after rewrite; "
                        f"annotate the BC with 'passthrough: true' to opt out",
                    )
                bc["passthrough"] = True
            bc["value"] = _serialize_expression(final)
            rewritten_via_bc_rule = True

    if not rewritten_via_bc_rule and value_raw is not None:
        bc["value"] = _rewrite_or_passthrough(
            f"{path}.value",
            value_raw,
            rules,
            ctx,
            max_passes,
            strict_unrewritten,
            passthrough,
            _set_passthrough,
        )


def _rewrite_or_passthrough(
    path: str,
    value_raw: Any,
    rules: Sequence[Rule],
    ctx: RuleContext,
    max_passes: int,
    strict_unrewritten: bool,
    passthrough: bool,
    set_passthrough: Callable[[bool], None],
) -> Any:
    expr = _parse_expression(value_raw)
    canon0 = canonicalize(expr)
    rewritten = canon0 if not rules else rewrite(canon0, rules, ctx, max_passes=max_passes)
    canon1 = canonicalize(rewritten)
    if passthrough:
        return _serialize_expression(canon1)
    if _has_pde_op(canon1):
        if strict_unrewritten:
            op = _first_pde_op(canon1)
            raise RuleEngineError(
                "E_UNREWRITTEN_PDE_OP",
                f"{path} still contains PDE op '{op}' after rewrite; "
                f"annotate the equation/BC with 'passthrough: true' to opt out",
            )
        set_passthrough(True)
    return _serialize_expression(canon1)


def _canonicalize_value(value_raw: Any) -> Any:
    expr = _parse_expression(value_raw)
    return _serialize_expression(canonicalize(expr))


# ============================================================================
# Leftover-PDE-op scan (RFC §11 Step 7)
# ============================================================================


def _has_pde_op(e: Expr) -> bool:
    return _first_pde_op(e) is not None


def _first_pde_op(e: Expr) -> Optional[str]:
    if isinstance(e, ExprNode):
        if e.op in _PDE_OPS:
            return e.op
        for a in e.args:
            r = _first_pde_op(a)
            if r is not None:
                return r
    return None


# ============================================================================
# Rule loading (permissive: accept array form or keyed-object form)
# ============================================================================


def _load_rules(raw: Any) -> List[Rule]:
    if raw is None:
        return []
    if not isinstance(raw, (list, Mapping)):
        return []
    if not raw:
        return []
    return parse_rules(raw)


# ============================================================================
# Misc helpers
# ============================================================================


def _as_bool(x: Any) -> bool:
    if isinstance(x, bool):
        return x
    if isinstance(x, str):
        return x.lower() == "true"
    if x is None:
        return False
    return x is True


# ============================================================================
# Metadata: discretized_from provenance
# ============================================================================


def _record_discretized_from(esm: Dict[str, Any]) -> None:
    meta_raw = esm.get("metadata")
    if isinstance(meta_raw, dict):
        meta = meta_raw
    elif isinstance(meta_raw, Mapping):
        meta = dict(meta_raw)
    else:
        meta = {}

    src_name = meta.get("name")
    provenance: Dict[str, Any] = {}
    if src_name is not None:
        provenance["name"] = str(src_name)
    meta["discretized_from"] = provenance

    tags = meta.get("tags")
    if isinstance(tags, list):
        if "discretized" not in (str(t) for t in tags):
            tags.append("discretized")
    else:
        meta["tags"] = ["discretized"]

    esm["metadata"] = meta

File: packages/esm-format-go/pkg/esm/discretize.go
// Discretization pipeline per discretization RFC §11 (gt-gbs2).
//
// Mirrors the Julia reference implementation in
// packages/EarthSciSerialization.jl/src/discretize.jl. DAE classification
// and the RFC §12 binding contract are out of scope for this binding and
// tracked separately.
//
// The public entry point is [Discretize] which walks a parsed ESM
// document and emits a discretized ESM:
//
//  1. Canonicalize all expressions (§5.4).
//  2. Resolve model-level boundary conditions into a synthetic `bc` op so
//     they flow through the same rule engine as interior equations.
//  3. Apply the rule engine (§5.2) to every equation RHS and every BC
//     value with a max-pass budget.
//  4. Re-canonicalize the rewritten ASTs.
//  5. Check for unrewritten PDE ops (§11 Step 7) — error or
//     passthrough-annotate depending on StrictUnrewritten.
//  6. Record `metadata.discretized_from` provenance.

package esm

import (
	"bytes"
	"encoding/json"
	"fmt"
)

// DiscretizeOptions configures the §11 pipeline.
type DiscretizeOptions struct {
	// MaxPasses is the per-expression rule-engine budget (§5.2.5). A
	// non-positive value uses DefaultMaxPasses.
	MaxPasses int
	// StrictUnrewritten controls what happens when a rewritten expression
	// still carries a PDE op: true (default) raises
	// E_UNREWRITTEN_PDE_OP; false stamps `passthrough: true` on the
	// offending equation/BC and retains it verbatim.
	StrictUnrewritten bool
}

// DefaultDiscretizeOptions returns the default pipeline configuration:
// max_passes=DefaultMaxPasses, strict_unrewritten=true.
func DefaultDiscretizeOptions() DiscretizeOptions {
	return DiscretizeOptions{
		MaxPasses:         DefaultMaxPasses,
		StrictUnrewritten: true,
	}
}

// Discretize runs the RFC §11 discretization pipeline on a parsed ESM
// document represented as a generic JSON map. Returns a new map; the
// input is not mutated.
//
// `esm` is the parsed ESM payload produced by JSON decoding (use
// encoding/json with UseNumber, or the lossless [LoadAsMap] helper).
// Errors are *RuleEngineError with stable codes per RFC §5.2 / §11.
func Discretize(esm map[string]interface{}, opts DiscretizeOptions) (map[string]interface{}, error) {
	if opts.MaxPasses <= 0 {
		opts.MaxPasses = DefaultMaxPasses
	}

	out := deepNativeMap(esm)

	topRules, err := loadRulesFromAny(out["rules"])
	if err != nil {
		return nil, err
	}

	ctx := buildDiscretizeContext(out)

	if models, ok := out["models"].(map[string]interface{}); ok {
		for mname, mraw := range models {
			model, ok := mraw.(map[string]interface{})
			if !ok {
				continue
			}
			if err := discretizeModel(mname, model, topRules, ctx, opts); err != nil {
				return nil, err
			}
			models[mname] = model
		}
	}

	recordDiscretizedFrom(out)
	return out, nil
}

// ============================================================================
// Rule context assembly (grids + variables)
// ============================================================================

func buildDiscretizeContext(esm map[string]interface{}) RuleContext {
	ctx := NewRuleContext()

	if gridsRaw, ok := esm["grids"].(map[string]interface{}); ok {
		for gname, graw := range gridsRaw {
			ctx.Grids[gname] = extractGridMeta(graw)
		}
	}

	if models, ok := esm["models"].(map[string]interface{}); ok {
		for _, mraw := range models {
			mobj, ok := mraw.(map[string]interface{})
			if !ok {
				continue
			}
			mgrid := stringOrEmpty(mobj["grid"])
			vars, ok := mobj["variables"].(map[string]interface{})
			if !ok {
				continue
			}
			for vname, vraw := range vars {
				vobj, ok := vraw.(map[string]interface{})
				if !ok {
					continue
				}
				meta := VariableMeta{Grid: mgrid}
				if shape, ok := vobj["shape"].([]interface{}); ok {
					ss := make([]string, 0, len(shape))
					for _, s := range shape {
						if str, ok := s.(string); ok {
							ss = append(ss, str)
						}
					}
					meta.Shape = ss
					meta.HasShape = true
				}
				if loc, ok := vobj["location"].(string); ok {
					meta.Location = loc
				}
				ctx.Variables[vname] = meta
			}
		}
	}

	return ctx
}

func extractGridMeta(graw interface{}) GridMeta {
	meta := GridMeta{}
	gobj, ok := graw.(map[string]interface{})
	if !ok {
		return meta
	}
	dims, ok := gobj["dimensions"].([]interface{})
	if !ok {
		return meta
	}
	for _, d := range dims {
		dobj, ok := d.(map[string]interface{})
		if !ok {
			continue
		}
		name, _ := dobj["name"].(string)
		if name == "" {
			continue
		}
		meta.SpatialDims = append(meta.SpatialDims, name)
		if p, ok := dobj["periodic"].(bool); ok && p {
			meta.PeriodicDims = append(meta.PeriodicDims, name)
		}
		if sp, ok := dobj["spacing"].(string); ok && (sp == "nonuniform" || sp == "stretched") {
			meta.NonuniformDims = append(meta.NonuniformDims, name)
		}
	}
	return meta
}

// ============================================================================
// Model-level pipeline
// ============================================================================

func discretizeModel(mname string, model map[string]interface{},
	topRules []Rule, ctx RuleContext, opts DiscretizeOptions) error {

	localRules, err := loadRulesFromAny(model["rules"])
	if err != nil {
		return err
	}
	var rules []Rule
	if len(localRules) == 0 {
		rules = topRules
	} else {
		rules = make([]Rule, 0, len(topRules)+len(localRules))
		rules = append(rules, topRules...)
		rules = append(rules, localRules...)
	}

	mp := lookupMaxPasses(model, opts.MaxPasses)

	if eqns, ok := model["equations"].([]interface{}); ok {
		for i, eqnAny := range eqns {
			eqn, ok := eqnAny.(map[string]interface{})
			if !ok {
				continue
			}
			path := fmt.Sprintf("models.%s.equations[%d]", mname, i)
			if err := discretizeEquation(path, eqn, rules, ctx, mp, opts.StrictUnrewritten); err != nil {
				return err
			}
			eqns[i] = eqn
		}
	}

	if bcs, ok := model["boundary_conditions"].(map[string]interface{}); ok {
		for bcName, bcAny := range bcs {
			bc, ok := bcAny.(map[string]interface{})
			if !ok {
				continue
			}
			path := fmt.Sprintf("models.%s.boundary_conditions.%s", mname, bcName)
			if err := discretizeBC(path, bc, rules, ctx, mp, opts.StrictUnrewritten); err != nil {
				return err
			}
			bcs[bcName] = bc
		}
	}
	return nil
}

func lookupMaxPasses(model map[string]interface{}, defaultMP int) int {
	rc, ok := model["rules_config"].(map[string]interface{})
	if !ok {
		return defaultMP
	}
	switch v := rc["max_passes"].(type) {
	case int:
		return v
	case int64:
		return int(v)
	case float64:
		if v == float64(int(v)) {
			return int(v)
		}
	case json.Number:
		if i, err := v.Int64(); err == nil {
			return int(i)
		}
	}
	return defaultMP
}

// ============================================================================
// Per-equation / per-BC rewrite
// ============================================================================

func discretizeEquation(path string, eqn map[string]interface{},
	rules []Rule, ctx RuleContext, maxPasses int, strict bool) error {

	passthrough := asBool(eqn["passthrough"])

	if rhsRaw, has := eqn["rhs"]; has {
		out, newPT, err := rewriteOrPassthrough(
			path+".rhs", rhsRaw, rules, ctx, maxPasses, strict, passthrough)
		if err != nil {
			return err
		}
		eqn["rhs"] = out
		if newPT && !passthrough {
			eqn["passthrough"] = true
		}
	}
	if lhsRaw, has := eqn["lhs"]; has {
		out, err := canonicalizeLHSValue(lhsRaw)
		if err != nil {
			return err
		}
		eqn["lhs"] = out
	}
	return nil
}

func discretizeBC(path string, bc map[string]interface{},
	rules []Rule, ctx RuleContext, maxPasses int, strict bool) error {

	passthrough := asBool(bc["passthrough"])
	variable := stringOrEmpty(bc["variable"])
	kind := stringOrEmpty(bc["kind"])
	side := stringOrEmpty(bc["side"])
	valueRaw, hasValue := bc["value"]

	// Step 1: try matching a `bc` rule pattern (§9.2 — synthetic wrapper).
	// The wrapper is `bc(variable[, value])`; `kind` / `side` are recorded
	// as sibling map fields for parity with the Julia reference but are
	// not consumed by the expression parser (rules match on op + args).
	rewrittenViaBCRule := false
	if variable != "" && kind != "" && len(rules) > 0 {
		wrapperMap := map[string]interface{}{
			"op":   "bc",
			"args": []interface{}{variable},
			"kind": kind,
		}
		if side != "" {
			wrapperMap["side"] = side
		}
		if hasValue {
			args := wrapperMap["args"].([]interface{})
			wrapperMap["args"] = append(args, valueRaw)
		}

		bcExpr, err := parseExprForRewrite(wrapperMap)
		if err != nil {
			return err
		}
		canonBC, err := Canonicalize(bcExpr)
		if err != nil {
			return err
		}
		rewriteOut, err := Rewrite(canonBC, rules, ctx, maxPasses)
		if err != nil {
			return err
		}
		node, isNode := asExprNode(rewriteOut)
		if !(isNode && node.Op == "bc") {
			final, err := Canonicalize(rewriteOut)
			if err != nil {
				return err
			}
			if hasPDEOp(final) && !passthrough {
				if strict {
					op := firstPDEOp(final)
					return newRuleErr("E_UNREWRITTEN_PDE_OP",
						fmt.Sprintf("%s.value still contains PDE op '%s' after rewrite; "+
							"annotate the BC with 'passthrough: true' to opt out", path, op))
				}
				bc["passthrough"] = true
			}
			bc["value"] = serializeExprToJSON(final)
			rewrittenViaBCRule = true
		}
	}

	// Step 2: default path — canonicalize `value` and run the rule engine.
	if !rewrittenViaBCRule && hasValue {
		out, newPT, err := rewriteOrPassthrough(
			path+".value", valueRaw, rules, ctx, maxPasses, strict, passthrough)
		if err != nil {
			return err
		}
		bc["value"] = out
		if newPT && !passthrough {
			bc["passthrough"] = true
		}
	}
	return nil
}

// rewriteOrPassthrough canonicalizes + rewrites + re-canonicalizes an
// expression. Returns (jsonValue, needsPassthroughStamp, err).
func rewriteOrPassthrough(path string, valueRaw interface{}, rules []Rule,
	ctx RuleContext, maxPasses int, strict, passthrough bool) (interface{}, bool, error) {

	expr, err := parseExprForRewrite(valueRaw)
	if err != nil {
		return nil, false, err
	}
	canon0, err := Canonicalize(expr)
	if err != nil {
		return nil, false, err
	}
	var rewritten Expression
	if len(rules) == 0 {
		rewritten = canon0
	} else {
		rewritten, err = Rewrite(canon0, rules, ctx, maxPasses)
		if err != nil {
			return nil, false, err
		}
	}
	canon1, err := Canonicalize(rewritten)
	if err != nil {
		return nil, false, err
	}
	if passthrough {
		return serializeExprToJSON(canon1), false, nil
	}
	if hasPDEOp(canon1) {
		if strict {
			op := firstPDEOp(canon1)
			return nil, false, newRuleErr("E_UNREWRITTEN_PDE_OP",
				fmt.Sprintf("%s still contains PDE op '%s' after rewrite; "+
					"annotate the equation/BC with 'passthrough: true' to opt out", path, op))
		}
		return serializeExprToJSON(canon1), true, nil
	}
	return serializeExprToJSON(canon1), false, nil
}

func canonicalizeLHSValue(raw interface{}) (interface{}, error) {
	expr, err := parseExprForRewrite(raw)
	if err != nil {
		return nil, err
	}
	c, err := Canonicalize(expr)
	if err != nil {
		return nil, err
	}
	return serializeExprToJSON(c), nil
}

// ============================================================================
// Parse/serialize bridges between JSON-tree form and Expression AST
// ============================================================================

// parseExprForRewrite converts a JSON-decoded value (map/slice/primitive) to
// an Expression AST suitable for Canonicalize / Rewrite.
func parseExprForRewrite(v interface{}) (Expression, error) {
	return parseExprValue(normalizeExprInput(v))
}

// normalizeExprInput normalizes json.Number tokens inside a decoded tree so
// that parseExprValue sees int64/float64 per §5.4.6. It returns a structurally
// identical tree (maps/slices are mutated in place; primitives converted).
func normalizeExprInput(v interface{}) interface{} {
	switch x := v.(type) {
	case json.Number:
		return normalizeJSONNumber(x)
	case map[string]interface{}:
		out := make(map[string]interface{}, len(x))
		for k, val := range x {
			out[k] = normalizeExprInput(val)
		}
		return out
	case []interface{}:
		out := make([]interface{}, len(x))
		for i, e := range x {
			out[i] = normalizeExprInput(e)
		}
		return out
	}
	return v
}

// serializeExprToJSON converts a canonicalized Expression back to a
// JSON-tree value (primitives + map[string]interface{} for ExprNode).
// The result round-trips through encoding/json and the canonical emitter.
func serializeExprToJSON(e Expression) interface{} {
	switch v := e.(type) {
	case nil:
		return nil
	case int64, float64, string, bool:
		return v
	case int:
		return int64(v)
	case ExprNode:
		return exprNodeToJSON(v)
	case *ExprNode:
		return exprNodeToJSON(*v)
	}
	return e
}

func exprNodeToJSON(n ExprNode) map[string]interface{} {
	out := map[string]interface{}{
		"op": n.Op,
	}
	args := make([]interface{}, len(n.Args))
	for i, a := range n.Args {
		args[i] = serializeExprToJSON(a)
	}
	out["args"] = args
	if n.Wrt != nil {
		out["wrt"] = *n.Wrt
	}
	if n.Dim != nil {
		out["dim"] = *n.Dim
	}
	if n.Name != nil {
		out["name"] = *n.Name
	}
	if n.Value != nil {
		out["value"] = n.Value
	}
	return out
}

// ============================================================================
// PDE-op scan (§11 Step 7)
// ============================================================================

var discretizePDEOps = map[string]struct{}{
	"grad": {}, "div": {}, "laplacian": {}, "D": {}, "bc": {},
}

func hasPDEOp(e Expression) bool {
	return firstPDEOp(e) != ""
}

func firstPDEOp(e Expression) string {
	node, ok := asExprNode(e)
	if !ok {
		return ""
	}
	if _, found := discretizePDEOps[node.Op]; found {
		return node.Op
	}
	for _, a := range node.Args {
		if op := firstPDEOp(a); op != "" {
			return op
		}
	}
	return ""
}

// ============================================================================
// Rules loading
// ============================================================================

func loadRulesFromAny(raw interface{}) ([]Rule, error) {
	if raw == nil {
		return nil, nil
	}
	switch raw.(type) {
	case []interface{}, map[string]interface{}:
		rules, err := parseRulesValue(normalizeExprInput(raw))
		if err != nil {
			return nil, err
		}
		return rules, nil
	}
	return nil, nil
}

// ============================================================================
// Metadata provenance
// ============================================================================

func recordDiscretizedFrom(esm map[string]interface{}) {
	meta, ok := esm["metadata"].(map[string]interface{})
	if !ok {
		meta = map[string]interface{}{}
	}
	provenance := map[string]interface{}{}
	if name, ok := meta["name"].(string); ok {
		provenance["name"] = name
	}
	meta["discretized_from"] = provenance

	switch tags := meta["tags"].(type) {
	case []interface{}:
		seen := false
		for _, t := range tags {
			if s, ok := t.(string); ok && s == "discretized" {
				seen = true
				break
			}
		}
		if !seen {
			meta["tags"] = append(tags, "discretized")
		}
	default:
		meta["tags"] = []interface{}{"discretized"}
	}

	esm["metadata"] = meta
}

// ============================================================================
// Deep copy + small helpers
// ============================================================================

// deepNativeMap returns a deep copy of a JSON-shaped map with numeric
// literals normalized to int64 / float64 per §5.4.6.
func deepNativeMap(m map[string]interface{}) map[string]interface{} {
	out := make(map[string]interface{}, len(m))
	for k, v := range m {
		out[k] = deepNative(v)
	}
	return out
}

func deepNative(v interface{}) interface{} {
	switch x := v.(type) {
	case nil:
		return nil
	case map[string]interface{}:
		out := make(map[string]interface{}, len(x))
		for k, val := range x {
			out[k] = deepNative(val)
		}
		return out
	case []interface{}:
		out := make([]interface{}, len(x))
		for i, e := range x {
			out[i] = deepNative(e)
		}
		return out
	case json.Number:
		return normalizeJSONNumber(x)
	}
	return v
}

func stringOrEmpty(v interface{}) string {
	s, _ := v.(string)
	return s
}

func asBool(v interface{}) bool {
	switch x := v.(type) {
	case bool:
		return x
	case string:
		return x == "true" || x == "True" || x == "TRUE"
	}
	return false
}

// ============================================================================
// JSON helpers for top-level document ingestion
// ============================================================================

// CanonicalDocJSON emits a discretized ESM document (or any JSON-tree value)
// as canonical JSON per RFC §5.4.6: sorted object keys at every level,
// minified (no inter-token whitespace), and canonical float literals
// (trailing `.0` for integer-valued magnitudes, lowercase `e` exponents).
func CanonicalDocJSON(doc interface{}) ([]byte, error) {
	return marshalCanonical(doc, false)
}

// LoadAsMap parses an ESM JSON document as a generic map preserving
// numeric int/float distinctions per §5.4.6. This is the preferred input
// form for Discretize since it does not require the strongly-typed
// EsmFile struct to understand every field (notably `rules`).
func LoadAsMap(data []byte) (map[string]interface{}, error) {
	dec := json.NewDecoder(bytes.NewReader(data))
	dec.UseNumber()
	var v interface{}
	if err := dec.Decode(&v); err != nil {
		return nil, fmt.Errorf("ESM decode: %w", err)
	}
	m, ok := v.(map[string]interface{})
	if !ok {
		return nil, fmt.Errorf("ESM decode: top-level value is not a JSON object")
	}
	return deepNativeMap(m), nil
}
